enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,442 @@
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annoated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contanis the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+3. OpenShift should create the OIDC provider in IAM
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seemsless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
+   machine-api) to use bound tokens
+
+## Proposal
+
+A new `aws-pod-identity-webhook` image will be added to the release payload
+based on [this](https://github.com/openshift/aws-pod-identity-webhook) repo.
+
+Considerable functionality will be added to the `cloud-credential-operator`
+(CCO).  This functionality is AWS-only and will only be active for clusters
+deployed with `platform: aws`.
+
+- Management of the `aws-pod-identity-webhook` and its dependent resources
+  - Synchronization of the `spec.clientConfig.caBundle` on
+    `MutationWebhookConfiguration` whenever the service serving CA
+    (`service-ca`) rotates
+- Creation of the OIDC provider in IAM (pending on Open Question)
+- Synchronization of the `keys.json` in the S3 bucket pointed to by the OIDC
+  provider with the rotatable public keys contained in
+  `bound-sa-token-signing-certs` configmap in the `openshift-kube-apiserver`
+  namespace (pending on Open Question)
+
+IAM permissions will need to be expanded to include IAM OIDC provider create and
+S3 bucket create/write if the CCO is to manage the OIDC provider (see Open
+Questions section)
+
+In terms of API changes, an additional `serviceAccountRef` field will be added
+to the `CredentialsRequests` CR `spec` alongside the existing [`secretRef`
+field](https://github.com/openshift/cloud-credential-operator/blob/9cc1c1abf898cfa08429d465da0f02d064bd89ee/manifests/00_v1_crd.yaml#L37-L42).
+Usage of this field is an indication by the user that they wish to use this new
+bound token injection mechanism.  The CCO will create a role, instead of a user,
+in response and create a `ServiceAccount` with the appropriate role annotation.
```

Comment:
> create a `ServiceAccount` with the appropriate role annotation

so if the service account needs to be created by the CCO, that means the namespace for the workload needs to pre-exist. How does that work with OLM/Helm installed operators.

EKS's documnented workflow separated the service account linking and setting up the IAM roles/users. therefore one could pre setup the cred request, and later on the installation created the service account with annotation allowing for mutation.

I think a user workflow example if included will help clearly define the expected flow.

Reply:
Yes, it does mean that the namespace must pre-exist.  This is the same for the secret based injection.  it would be nice if we required the `CredentialsRequest` CR be created in same namespace the injected secret/serviceaccount, but... that is not what we did.

Ideally, the flow is the same from the user perspective.  Only the injection method changes.  Pods that expect injected credentials are blocked from starting because their serviceaccount doesn't exist just like they don't start for a missing secret.

Only difference in this flow is that the underlying steps can be done in different orders.  The ordering of the serviceaccount creation and role creation is doesn't matter as long as they are both done by the time the pod is created and the webhook has to create the linkage.

Reply:
> Yes, it does mean that the namespace must pre-exist. This is the same for the secret based injection. it would be nice if we required the `CredentialsRequest` CR be created in same namespace the injected secret/serviceaccount, but... that is not what we did.

@sjenning when you say "that is not what did", can you elaborate what code you are looking at when you say that?

thanks

@bparees ^^

> 
> Ideally, the flow is the same from the user perspective. Only the injection method changes. Pods that expect injected credentials are blocked from starting because their serviceaccount doesn't exist just like they don't start for a missing secret.
> 
> Only difference in this flow is that the underlying steps can be done in different orders. The ordering of the serviceaccount creation and role creation is doesn't matter as long as they are both done by the time the pod is created and the webhook has to create the linkage.

Reply:
@gabemontero what I mean is that we should have enforced the namespace of the destination secret match the namespace of the CredentialsRequest CR i.e. you create the CR in the namespace you want the secret injected.  IIRC all the CRs are created in the CCO namespace, right?

Reply:
just verified that all `CredentialsRequest` CRs are created in the `openshift-cloud-credential-operator` namespace.  that was kinda a weird design choice but... done now.  Seems like creating the CRs in the namespace in which the secret/serviceaccount ref was going to be created would have been better.

Reply:
@abhinavdahiya the flow will be identical to EKS https://github.com/openshift/aws-pod-identity-webhook#eks-walkthrough

EKS runs the webhook on their service side, not visible to the user, but the `MutatingWebhookConfiguration` _is_ in the user cluster and created automatically.  It is just that the Role session can't be created until the user registers the ODIC provider (endpoint provided by EKS) in IAM since AWS will have no way to validate the tokens from the cluster.

Reply:
It's been too long and I can't find any history as to why we started dropping all cred requests in the CCO namespace, IIRC something to do with chicken/egg release image scenarios when the namespace can't exist yet, or something to do with RBAC and namespace admins. I believe that today CCO looks like it will react to a CR in any namespace, but there's a comment that we should revisit that.

CCO watches namespaces so it can act immediately once a target namespace comes into existence and that could be re-used here.

Reply:
@gabemontero might be some crossing of the streams here.

This enhancement enables the injection of AWS STS bounded tokens into pods so they can assume IAM roles.  The RH entitlement credential injection is platform independent and a different thing altogether.

Reply:
certainly the overall EPs are apples and oranges @sjenning 

that doesn't mean there couldn't be some common sub parts ... that is all I was exploring

but we don't have in this case

and thanks for indulging me